snapshot/volume-snapshotting-proposal.md

```diff
@@ -0,0 +1,412 @@
+Kubernetes Snapshotting Proposal
+================================
+
+**Authors:** [Cindy Wang](https://github.com/ciwang), [Jing Xu](https://github.com/jinxu97),[Tomas Smetana](https://github.com/tsmetana), [Huamin Chen ](https://github.com/rootfs)
+
+## Background
+
+Many storage systems (GCE PD, Amazon EBS, etc.) provide the ability to create "snapshots" of a persistent volumes to protect against data loss. Snapshots can be used in place of a traditional backup system to back up and restore primary and critical data. Snapshots allow for quick data backup (for example, it takes a fraction of a second to create a GCE PD snapshot) and offer fast recovery time objectives (RTOs) and recovery point objectives (RPOs).
+
+Typical existing backup solutions offer on demand or scheduled snapshots.
+
+An application developer using a storage may want to create a snapshot before an update or other major event. Kubernetes does not currently offer a standardized snapshot API for creating, listing, deleting, and restoring snapshots on an arbitrary volume.
+
+Existing solutions for scheduled snapshotting include [cron jobs](https://forums.aws.amazon.com/message.jspa?messageID=570265) and [external storage drivers](http://rancher.com/introducing-convoy-a-docker-volume-driver-for-backup-and-recovery-of-persistent-data/). Some cloud storage volumes can be configured to take automatic snapshots, but this is specified on the volumes themselves.
+
+## Objectives
+
+For the first version of snapshotting support in Kubernetes, only on-demand snapshots will be supported. Features listed in the roadmap for future versions are also nongoals.
+The initial version is created as an external controller and provisioner (i.e. two separate binaries).
+
+* Goal 1: Enable *on-demand* snapshots of Kubernetes persistent volumes by application developers.
+
+    * Nongoal: Enable *automatic* periodic snapshotting for direct volumes in pods.
+
+* Goal 2: Expose standardized snapshotting operations to create, list and delete snapshots in Kubernetes REST API.
+
+* Goal 3: Implement snapshotting interface for Amazon EBS.
+
+* Goal 4: Implement snapshotting interface for GCE PDs.
+
+### Feature Roadmap
+
+Major features, planned for the first version:
+
+* On demand snapshots
+
+    * API to create new snapshots
+
+    * API to list snapshots available to the user
+
+    * API to delete existing snapshots
+
+    * API to create a persistent volume from a snapshot
+
+### Future Features
+
+The features that are not planned for the first version of the API bus should be considered in future versions:
+
+* Creating snapshots
+
+    * Scheduled and periodic snapshots
+
+    * Aplication initiated on-demand snapshot creation
+
+    * Support snapshot per PVC, pod or StatefulSet
+
+    * Support snapshots for non-cloud storage volumes (i.e. plugins that require actions to be triggered from the node)
```

Comment:
May be this can be better worded. That said,  non-cloud storage volumes are not clear to me. Thinking specificically about GlusterFS, it dont require any actions to be triggered from the kubelet.